kafka-rest/src/main/java/io/confluent/kafkarest/controllers/AbstractConsumerLagManager.java

```diff
@@ -0,0 +1,158 @@
+import org.apache.kafka.clients.admin.OffsetSpec;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.IsolationLevel;
+import org.apache.kafka.common.TopicPartition;
+
+abstract class AbstractConsumerLagManager {
+
+  private final Admin kafkaAdminClient;
+
+  @Inject
+  AbstractConsumerLagManager(
+      Admin kafkaAdminClient
+  ) {
+    this.kafkaAdminClient = kafkaAdminClient;
+  }
+
+  protected final Map<TopicPartition, MemberId> getMemberIds(
+      ConsumerGroup consumerGroup
+  ) {
+    Map<TopicPartition, MemberId> tpMemberIds = new HashMap<>();
+    for (Consumer consumer: consumerGroup.getConsumers()) {
+      for (Partition partition : consumer.getAssignedPartitions()) {
+        MemberId memberId =
+            MemberId.builder()
+                .setConsumerId(consumer.getConsumerId())
+                .setClientId(consumer.getClientId())
+                .setInstanceId(consumer.getInstanceId().orElse(null))
+                .build();
+        TopicPartition topicPartition =
+            new TopicPartition(partition.getTopicName(), partition.getPartitionId());
+        tpMemberIds.put(topicPartition, memberId);
+      }
+    }
+    return tpMemberIds;
+  }
+
+  protected final CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> getCurrentOffsets(
+      String consumerGroupId
+  ) {
+    return KafkaFutures.toCompletableFuture(
+        kafkaAdminClient.listConsumerGroupOffsets(
+            consumerGroupId, new ListConsumerGroupOffsetsOptions())
+            .partitionsToOffsetAndMetadata());
+  }
+
+  protected final CompletableFuture<Map<TopicPartition, ListOffsetsResultInfo>> getLatestOffsets(
+      Map<TopicPartition, OffsetAndMetadata> currentOffsets
+  ) {
+    IsolationLevel isolationLevel = IsolationLevel.READ_COMMITTED;
+    Map<TopicPartition, OffsetSpec> latestOffsetSpecs =
+        currentOffsets.keySet()
+            .stream()
+            .collect(
+                Collectors.toMap(Function.identity(), topicPartition -> OffsetSpec.latest()));
+    return KafkaFutures.toCompletableFuture(
+        kafkaAdminClient.listOffsets(
+            latestOffsetSpecs,
+            new ListOffsetsOptions(isolationLevel)).all());
+  }
+
+  protected final long getCurrentOffset(Map<TopicPartition, OffsetAndMetadata> map, TopicPartition topicPartition) {
+    if (map == null) {
+      return -1;
+    }
+    OffsetAndMetadata oam = map.get(topicPartition);
+    if (oam == null) {
+      return -1;
+    }
+    return oam.offset();
+  }
+
+  protected final long getOffset(Map<TopicPartition, ListOffsetsResultInfo> map, TopicPartition topicPartition) {
+    if (map == null) {
+      return -1;
+    }
+
+    ListOffsetsResultInfo offsetInfo = map.get(topicPartition);
+    if (offsetInfo == null) {
+      return -1;
+    }
+
+    return offsetInfo.offset();
+  }
+
+  protected final <T extends Map> T checkOffsetsExist(T offsets, String message, Object... args) {
```

Comment:
`<T extends Map>` is probably generating a warning in your IDE, which means you probably should not be doing it :). Use `<T extends Map<?, ?>>` instead.